calico.go

```diff
@@ -0,0 +1,302 @@
+			theEndpoint.Mac = "EE:EE:EE:EE:EE:EE"
+			theEndpoint.State = "active"
+
+			if result.IP4 != nil {
+				result.IP4.IP.Mask = net.CIDRMask(32, 32)
+				theEndpoint.IPv4Nets = []string{result.IP4.IP.String()}
+			} else {
+				theEndpoint.IPv4Nets = []string{}
+			}
+
+			if result.IP6 != nil {
+				result.IP6.IP.Mask = net.CIDRMask(128, 128)
+				theEndpoint.IPv6Nets = []string{result.IP6.IP.String()}
+			} else {
+				theEndpoint.IPv6Nets = []string{}
+			}
+			fmt.Fprintf(os.Stderr, "Calico CNI using IPv4=%s IPv6=%s\n", theEndpoint.IPv4Nets, theEndpoint.IPv6Nets)
+
+			// 3) Set up the veth
+			hostVethName, contVethMac, err := DoNetworking(args, conf, result)
+			if err != nil {
+				return err
+			}
+			theEndpoint.Mac = contVethMac
+			theEndpoint.Name = hostVethName
+		}
+
+		// Write the endpoint object (either the newly created one, or the updated one with a new ProfileIDs).
+		if err := theEndpoint.Write(etcd); err != nil {
+			return err
+		}
+	}
+
+	// Handle profile creation.
+	// If Kubernetes is being used then profiles only need to be created if there is no policy block in the network
+	// config. If there is a policy block then "proper" policy is being used and the policy controller handles
+	// profile creation.
+	if !runningUnderK8s || conf.Policy == nil {
+		//TODO - this is the wrong test. It should be on the policy type
+		// Start by checking if the profile already exists. If it already exists then there is no work to do (the CNI plugin never updates a profile).
+		exists, err := libcalico.ProfileExists(conf.Name, etcd)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			// The profile doesn't exist so needs to be created. The rules vary depending on whether k8s is being used.
+			// Under k8s (without full policy support) the rule is permissive and allows all traffic.
+			// Otherwise, incoming traffic is only allowed from profiles with the same tag.
+			k8sInboundRule := []libcalico.Rule{libcalico.Rule{Action: "allow"}}
+			tagInboundRule := []libcalico.Rule{libcalico.Rule{Action: "allow", SrcTag: conf.Name}}
+			fmt.Fprintf(os.Stderr, "Calico CNI creating profile. profile=%s\n", conf.Name)
+
+			var inboundRule []libcalico.Rule
+			if runningUnderK8s {
+				inboundRule = k8sInboundRule
+			} else {
+				inboundRule = tagInboundRule
+			}
+
+			profile := libcalico.Profile{
+				ID: conf.Name,
+				Rules: libcalico.Rules{
+					Inbound:  inboundRule,
+					Outbound: []libcalico.Rule{libcalico.Rule{Action: "allow"}}},
+				Tags: []string{conf.Name}}
+			if err := profile.Write(etcd); err != nil {
+				return err
+			}
+		}
+	}
+
+	return result.Print()
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	n := NetConf{}
+	if err := json.Unmarshal(args.StdinData, &n); err != nil {
+		return fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	if err := AddIgnoreUnknownArgs(); err != nil {
+		return err
+	}
+
+	// Always try to release the address
+	if err := ipam.ExecDel(n.IPAM.Type, args.StdinData); err != nil {
```

Comment:
We're short-circuiting if we encounter an error releasing the IP address.  Is this OK?

I _think_ we want to hold on to this error, and attempt to cleanup the endpoint as well.  Is there a way to return/combine multiple errors somehow?